opentelemetry-api/src/opentelemetry/trace/span.py

```diff
@@ -157,26 +162,58 @@ class SpanContext:
         is_remote: True if propagated from a remote parent.
     """
 
-    def __init__(
-        self,
+    def __new__(
+        cls,
         trace_id: int,
         span_id: int,
         is_remote: bool,
         trace_flags: "TraceFlags" = DEFAULT_TRACE_OPTIONS,
         trace_state: "TraceState" = DEFAULT_TRACE_STATE,
-    ) -> None:
+    ) -> "SpanContext":
         if trace_flags is None:
             trace_flags = DEFAULT_TRACE_OPTIONS
         if trace_state is None:
             trace_state = DEFAULT_TRACE_STATE
-        self.trace_id = trace_id
-        self.span_id = span_id
-        self.trace_flags = trace_flags
-        self.trace_state = trace_state
-        self.is_remote = is_remote
-        self.is_valid = (
-            self.trace_id != INVALID_TRACE_ID
-            and self.span_id != INVALID_SPAN_ID
+
+        is_valid = trace_id != INVALID_TRACE_ID and span_id != INVALID_SPAN_ID
+
+        return tuple.__new__(
```

Comment:
Since we have to override setattr anyways is there an advantage of using tuple vs normal fields set with super.setattr? Latter might be a bit simpler.

Reply:
Hmm, if I understand correctly, I think the advantage of inheriting from a tuple is that it doesn't allow something like `object.__setattr__(context, "trace_id", 2)` (while not inheriting does). I'm not sure if that's something we want to avoid or if we should just keep it simple.

Reply:
Hmm, it actually seems a lot cleaner and I dont think the case I mentioned above is something that needs to be avoided.

Reply:
I see - in that case no worries :)

Reply:
> Hmm, if I understand correctly, I think the advantage of inheriting from a tuple is that it doesn't allow something like object.__setattr__(context, "trace_id", 2) (while not inheriting does). I'm not sure if that's something we want to avoid or if we should just keep it simple.

Do we want to make it impossible to use `__setattr__`? This should be immutable to prevent users from the mistake of using SpanContext to additional information around. I think making that impossible to do through the public API serves that purpose and self-documents the class. I don't think we need to make it literally impossible to modify the object. Given this is Python, I think anyone included to do it will find a way. We should just make sure we inform the person that they shouldn't do it. If they want to go out of their way and still modify, I think we should probably allow that.

I'm personally fine with using tuple for this but if it is causing too much trouble with tooling and needs a lot of other machinery, perhaps we could use a simpler way that protect modifications through public API but doesn't actually try to make it impossible for users to modify the object.

Reply:
@owais @anuraaga Hmm, I think I solved the tooling issues now (the ignores should be inline). There seems to be a false-positive for `pylint` in determining if SpanContext was instantiated as a tuple (here: https://github.com/open-telemetry/opentelemetry-python/pull/1134/files#diff-36ec0e77c55d7f4a4caef71eda0b3191R187). The issue for testing if we could change the attributes was also fixed inline (here: https://github.com/open-telemetry/opentelemetry-python/pull/1134/files#diff-0924a59286313e337f50e00112a27df8R46-R51). 

I'm not sure if it's worth it to go through with the change as it seems like Python doesn't support immutability very well, but all the tooling changes that I made globally were removed.